index.js

```diff
@@ -55,21 +55,54 @@ function findAllDependencies(file, knownDependencies, baseDir, knownFiles) {
     return knownDependencies;
   }
 
-  if (baseDir) {
-    return findAllDependenciesHelp(file, knownDependencies, baseDir, knownFiles).then(function(thing){
+  if (sourceDirectories) {
+    return findAllDependenciesHelp(file, knownDependencies, sourceDirectories, knownFiles).then(function(thing){
       return thing.knownDependencies;
     });
   } else {
-    return getBaseDir(file).then(function(newBaseDir) {
-      return findAllDependenciesHelp(file, knownDependencies, newBaseDir, knownFiles).then(function(thing){
-        return thing.knownDependencies;
+    return getBaseDir(file)
+      .then(findElmPackage)
+      .then(getSourceDirectories)
+      .then(function(newSourceDirs) {
+        return findAllDependenciesHelp(file, knownDependencies, newSourceDirs, knownFiles).then(function(thing){
+          return thing.knownDependencies;
+        });
       });
-    })
   }
 }
 
+// Given a source directory (containing top-level Elm modules), locate the
+// elm-package.json file that includes it.
+function findElmPackage(baseDir, currentDir) {
```

Comment:
Thinking about it some more, I’m thinking we probably should check that the `source-directories` does contain `baseDir`. If not, and the elm-package.json contains source directories that don’t match, find-elm-dependencies is going to be unable to find any dependencies in that `baseDir`, and it will be really confusing.

I’ll update this and include a test case.